Don't chain imports with *

esoergel opened the pull request:
From the slack convo yesterday.
@gcapalbo

esoergel committed:
Don't chain imports with *

`logging` and the django `User` model were both being imported from
`users/signals.py`, hopefully by accident.  This adds those explicitly
as dependencies, and gets rid of the `import *`.

esoergel committed:
Stick in another auto-import for shell_plus

czue commented on corehq/apps/users/models.py:
```diff
@@ -2548,7 +2550,3 @@ def get(self, user_id):
             return user
-
-
-from .signals import *
```
I think the (very long time ago) purpose of this import was to make sure that all the `signal.connect` code fired whenever django bootstrapped. haven't read slack convo yet, but wanted to confirm that we were confident this change didn't accidentally change that behavior

gcapalbo replied:
that's a good point, it looks like signals.py has some code it executes to setup post-save signals which now won't get setup until the import happens within the function calls above

czue replied:
it looks like we should replace this with an `AppConfig`. you can see @snopoke's recent change to fluff as an example: https://github.com/dimagi/fluff/pull/105

snopoke replied:
See [Where should this code live?](https://docs.djangoproject.com/en/1.8/topics/signals/#connecting-receiver-functions)

esoergel committed:
Register signals from the user AppConfig

esoergel committed:
Protect signals against duplication in tests

https://docs.djangoproject.com/en/1.8/topics/signals/#preventing-duplicate-signals

esoergel commented on corehq/apps/users/__init__.py:
```diff
@@ -1,4 +1,17 @@
+
+    def ready(self):
+        """Code to run with Django starts"""
+        ExtraPreindexPlugin.register('users', __file__, settings.NEW_USERS_GROUPS_DB)
```
@dannyroberts FYI - I took the liberty of moving this into `AppConfig.ready()`.  Know of any reason why that's not a good idea?

dannyroberts replied:
Off the top of my head, this looks fine.

esoergel commented on corehq/apps/users/__init__.py:
```diff
@@ -1,4 +1,17 @@
+        connect_user_signals()
+
+
+default_app_config = 'corehq.apps.users.UsersAppConfig'
```
According to the docs, you can just change the `INSTALLED_APPS` entry to point to this class, but I kept getting a weird, cryptic error when I tried that - I think we do other things with `INSTALLED_APPS`, and changing this broke something non-obvious.  This seems to work, though.

esoergel commented on corehq/apps/users/signals.py:
```diff
@@ -11,6 +11,10 @@
 
 
+commcare_user_post_save = Signal(providing_args=["couch_user"])
+couch_user_post_save = Signal(providing_args=["couch_user"])
```
It felt like these two should be right at the top of the file

gcapalbo commented:
:+1: Will give others a chance to review before merge as well, but looks good

Do you know if there's test coverage on the things that connect to those signals?

esoergel commented:
@gcapalbo I do not, but I verified that the signal handlers are being
called on save

On Tue, Oct 20, 2015 at 4:35 PM, gcapalbo <notifications@github.com> wrote:

> [image: :+1:] Will give others a chance to review before merge as well,
> but looks good
>
> Do you know if there's test coverage on the things that connect to those
> signals?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/8693#issuecomment-149694221>.
>



--

czue commented:
adding tests for each signal sounds like it'd be easy and helpful

snopoke commented on corehq/apps/users/models.py:
```diff
@@ -1414,7 +1416,7 @@ def clear_quickcache_for_user(self):
         super(CommCareUser, self).save(**params)
 
-        from corehq.apps.users.signals import commcare_user_post_save
+        from .signals import commcare_user_post_save
```
Why the change to relative import?

esoergel replied:
No particular reason - I was adding the same import before each of the signal connections, so I just made them consistent, and I generally prefer explicit relative imports for stuff within the same module.  Actually, I could probably just import this at the top of the file now that the signal binding has moved to the AppConfig.

esoergel committed:
Include error class on signal failure exceptions

esoergel committed:
Add `mock_out_couch` decorator

It's a very thin abstraction of FakeCouchDb, but I hate working with the
mock library directly, and it's always hard to remember what to mock, so
hopefully this will encourage more people to use it.

You can use this to decorate a `SimpleTestCase` test class, and it'll
mock db calls to return nothing.  You could specify actual things to
return, but we don't use that part anywhere yet.

esoergel committed:
Test CommCareUser.save signal handlers

esoergel committed:
Add tests for WebUser signals

esoergel commented:
Turns out testing signals was riddled with confusing gotchas.  I finally got it working though, and I think I understand the problems, so feel free to ping me if that's something you're trying to do in the future.  I also left some comments in the test file.

As a bonus, I added a [`FakeCouch` decorator](https://github.com/dimagi/commcare-hq/commit/4ec66e80ffdcb83306d516068ff4bad58d74201a)

gcapalbo commented:
Thanks @esoergel for putting that together.  I think the build failure is real though.

esoergel committed:
Don't try to hit postgres db.

We should probably figure out a good way to mock out calls to that too.

czue commented:
(kicked build)
